### PR TITLE
Makes audit math robust to ties and recounts

### LINKED ISSUE #812: Float division by 0 error on AB signoff

### PR DESCRIPTION
**Pull request by umbernhard**
**Description**

Updates the three implemented classes of audit math to handle tied elections and full hand recounts correctly. Fixes #812, starts to address #805. 

**Testing**
Adds new tests for tied election for each audit class. Updates BRAVO to handle ties more appropriately. 

**Progress**

Ready.

### CODE REVIEW

File: server/audit_math/macro.py [hunk 1]
```diff
@@ -57,6 +57,8 @@ def compute_error(
             a_lp = sampled_results[contest.name][loser]
 
             V_wl = contest.candidates[winner] - contest.candidates[loser]
+            if not V_wl:
```
Comment (jonahkagan) [inline server/audit_math/macro.py:60]: small nit: when dealing with numbers, i prefer explicit: `if V_wl == 0:`

File: server/audit_math/macro.py [hunk 2]
```diff
@@ -200,6 +205,9 @@ def get_sample_sizes(
 
     if not U:
         return 1
+    elif U < 0:
+        # This means we have a tie
+        return len(reported_results)
```
Comment (jonahkagan) [inline server/audit_math/macro.py:210]: interesting, so in a tie we do a full recount. i was thinking we'd just fail but i guess this is a more graceful answer

File: server/audit_math/macro.py [hunk 3]
```diff
@@ -240,6 +248,10 @@ def compute_risk(
     """
     assert risk_limit < 1, "The risk-limit must be less than one!"
 
+    # We've done a full hand recount
+    if len(sample_results) == len(reported_results):
+        return 0, True
```
Comment (jonahkagan) [inline server/audit_math/macro.py:253]: we don't get a p-value in a full hand recount? what if the recount results don't match the reported results?

File: server/audit_math/macro.py [hunk 4]
```diff
@@ -252,7 +264,10 @@ def compute_risk(
 
         taint = e_p / u_p
 
-        p *= (1 - 1 / U) / (1 - taint)
+        if taint == 1:
+            p *= 1
```
Comment (jonahkagan) [inline server/audit_math/macro.py:268]: ok maybe im missing something but what's the point of multiplying by 1?

File: server/audit_math/supersimple.py [hunk 5]
```diff
@@ -139,7 +139,13 @@ def compute_discrepancies(
                 if e:
                     # we found a discrepancy!
                     found = True
-                e_weighted = e / V_wl
+
+                if not V_wl:
+                    # In this case the error is undefined
+                    e_weighted = -1.0
```
Comment (jonahkagan) [inline server/audit_math/supersimple.py:145]: can you explain this case? i dont understand it. when would `V_wl` be 0 and why do you then set `e_weighted` to -1.0?

File: server/audit_math/supersimple.py [hunk 6]
```diff
@@ -270,11 +274,22 @@ def compute_risk(
...
 
-        denom = (2 * gamma) / V
-        p_b = (1 - 1 / U) / (1 - (e_r / denom))
+        if contest.diluted_margin:
+            U = 2 * gamma / contest.diluted_margin
+            denom = (2 * gamma) / V
+            p_b = (1 - 1 / U) / (1 - (e_r / denom))
+        else:
+            # If the contest is a tie, this step results in 1 - 1/(infinity)
+            # divided by 1 - e_r/infinity, i.e. 1
```
Comment (jonahkagan) [inline server/audit_math/supersimple.py:283]: had to think about this lol, but i think i agree!

File: server/tests/audit_math/test_bravo.py [hunk 7]
```diff
@@ -379,6 +379,44 @@ def test_compute_risk_empty(contests):
...
+
+    # Now do a full hand recount
+    sample_results = {
+        "cand1": 501,
+        "cand2": 499,
+    }
+
+    computed_p, res = bravo.compute_risk(RISK_LIMIT, contest, sample_results)
+
+    assert not computed_p[("cand1", "cand2")]
```
Comment (jonahkagan) [inline server/tests/audit_math/test_bravo.py:416]: similar nit as to above - are you asserting that it's `None` or `0`?

File: server/tests/ballot_comparison/snapshots/snap_test_ballot_comparison.py [hunk 8]
```diff
@@ -100,7 +100,7 @@
 Round Number,Contest Name,Targeted?,Sample Size,Risk Limit Met?,P-Value,Start Time,End Time,Audited Votes\r
 1,Contest 1,Targeted,22,No,0.1128631733,DATETIME,DATETIME,Choice 1-1: 5; Choice 1-2: 1\r
 1,Contest 2,Opportunistic,,No,0.1888821229,DATETIME,DATETIME,Choice 2-1: 4; Choice 2-2: 1; Choice 2-3: 0\r
-2,Contest 1,Targeted,41,Yes,0.0008560347,DATETIME,DATETIME,Choice 1-1: 16; Choice 1-2: 6\r
+2,Contest 1,Targeted,41,Yes,<0.0000000001,DATETIME,DATETIME,Choice 1-1: 16; Choice 1-2: 6\r
```
Comment (jonahkagan) [inline server/tests/ballot_comparison/snapshots/snap_test_ballot_comparison.py:103]: i know im always doing this, but do you have an idea of what changed here? i dont think this test case should have hit any of the edge cases this PR is accounting for

File: server/audit_math/macro.py (hunk 2 above)
Comment (umbernhard) [reply server/audit_math/macro.py:210]: Yeah. I don't expect Arlo will ever be used in this way (except for opportunistic contests, maybe?) but regardless it hedges against oddities like the funky CVRs @mcchilders was using.

File: server/audit_math/macro.py (hunk 3 above)
Comment (umbernhard) [reply server/audit_math/macro.py:253]: So in a recount p-values no longer mean much. If you've counted all of your ballots by hand, then it is necessarily the case that there is zero chance that a computer malfunction could have caused a miscount (since no computers were used to do the count). While we could compute a p-value, it would be a little like estimating the chance it is raining while it is actively raining.

File: server/audit_math/macro.py (hunk 4 above)
Comment (umbernhard) [reply server/audit_math/macro.py:268]: if the taint is one, then the p-value is undefined since the KM denominator goes to zero. I suppose the correct thing to do would be to use Decimal's `inf` value, because the p-value goes to infinity in this case. Essentially we're trying to say that if we ever observe a batch that obtains its maximum possible error, then something has truly gone horribly wrong and we need to do a full hand recount.

Comment (umbernhard) [reply server/audit_math/macro.py:268]: Multiplying p by 1 here is approximately equivalent to the p-value blowing up, but not exactly. I can change it appropriately.

File: server/audit_math/supersimple.py (hunk 5 above)
Comment (umbernhard) [reply server/audit_math/supersimple.py:145]: If V_wl is 0, the race was tied, meaning that the denominator of the weighted error is 0 and the weighted error approaches infinity. I'm using -1 here to code that this has occurred, so later on we can gracefully handle some of the math. I suppose I should pick a better value like `inf`.

File: server/tests/audit_math/test_bravo.py (hunk 7 above)
Comment (umbernhard) [reply server/tests/audit_math/test_bravo.py:416]: 0. I will change it to a `==`

File: server/tests/ballot_comparison/snapshots/snap_test_ballot_comparison.py (hunk 8 above)
Comment (umbernhard) [reply server/tests/ballot_comparison/snapshots/snap_test_ballot_comparison.py:103]: Ah, I think this may have been a corner case identified above where the weighted error was improperly handled. I'll investigate further.

File: server/audit_math/macro.py (hunk 3 above)
Comment (jonahkagan) [reply server/audit_math/macro.py:253]: might be nice to include that explanation as a comment for future readers

File: server/audit_math/macro.py (hunk 4 above)
Comment (jonahkagan) [reply server/audit_math/macro.py:268]: cool, that all makes sense to me except why multiplying by 1 blows up the p-value. in math, doesn't multiplying by 1 usually not change the number?

File: server/audit_math/supersimple.py (hunk 5 above)
Comment (jonahkagan) [reply server/audit_math/supersimple.py:145]: ah that makes sense

Comment (jonahkagan) [reply server/audit_math/supersimple.py:145]: agreed, inf would be better

File: server/tests/ballot_comparison/snapshots/snap_test_ballot_comparison.py (hunk 8 above)
Comment (umbernhard) [reply server/tests/ballot_comparison/snapshots/snap_test_ballot_comparison.py:103]: Revisiting this, this contest reports 24 ballots total but has 16 votes for the winner and 10 votes for the loser for a total of 26 votes. That should never happen. 

I dug into it some more, and it appears that the p-value really should be that low. I think the previous result was being affected by the precision bug!

File: server/tests/audit_math/test_supersimple.py [hunk 9]
```diff
@@ -297,6 +297,106 @@ def test_compute_risk(contests, cvrs):
...
+def test_snapshot_test():
+    contest_data = {
+        "winner": 16,
+        "loser": 10,
+        "ballots": 26,
+        "numWinners": 1,
+        "votesAllowed": 1,
+    }
+
+    contest = Contest("Jonah Test", contest_data)
```
Comment (jonahkagan) [inline server/tests/audit_math/test_supersimple.py:358]: lol i dont need to be immortalized

File: server/tests/ballot_comparison/snapshots/snap_test_ballot_comparison.py (hunk 8 above)
Comment (jonahkagan) [reply server/tests/ballot_comparison/snapshots/snap_test_ballot_comparison.py:103]: shouldn't this have changed with the precision PR then, not this PR?

File: Pipfile [hunk 10]
```diff
@@ -4,7 +4,7 @@ url = "https://pypi.org/simple"
 verify_ssl = true
 
 [dev-packages]
-mypy = "*"
+mypy = "==0.780"
```
Comment (jonahkagan) [inline Pipfile:7]: i upgraded to the latest mypy version on `main` and it didn't cause any problems. do you remember what happened when you tried?

Comment (jonahkagan) [reply Pipfile:7]: oh weird i made a new branch to check and it does fail on CI: https://app.circleci.com/pipelines/github/votingworks/arlo/3144/workflows/4410b01b-67c4-4273-8c44-1297e83a0f04/jobs/4966

Comment (umbernhard) [reply Pipfile:7]: It was complaining about types in `server/util/group_by.py` and somewhere else. See here: https://app.circleci.com/pipelines/github/votingworks/arlo/3137/workflows/3be6d301-e3cf-46c3-aa56-8eb77e560549/jobs/4936

File: server/audit_math/macro.py (hunk 4 above)
Comment (umbernhard) [reply server/audit_math/macro.py]: I fixed this by setting p to `Decimal('inf')`

File: server/tests/ballot_comparison/snapshots/snap_test_ballot_comparison.py (hunk 8 above)
Comment (umbernhard) [reply server/tests/ballot_comparison/snapshots/snap_test_ballot_comparison.py]: Oh, I see what's happening here. Because we have more ballots in the CVR than the `total_ballots` field, the audit thinks a full-hand recount has happened in the second round, so it sets the p value to 0.

Comment (umbernhard) [reply server/tests/ballot_comparison/snapshots/snap_test_ballot_comparison.py]: So yeah, this test case shouldn't really correspond to real-world, ever. We'd never sample 24 ballots 41 times :joy:

Comment (jonahkagan) [reply server/tests/ballot_comparison/snapshots/snap_test_ballot_comparison.py]: ok that makes sense. kind of surprised it didn't just print a p-value of 0 though. but maybe we don't expect that to actually happen so it doesn't matter?

File: Pipfile (hunk 10 above)
Comment (jonahkagan) [reply Pipfile:7]: fixing this on #849

Comment (jonahkagan) [reply Pipfile:7]: so you can take this out

File: server/tests/ballot_comparison/snapshots/snap_test_ballot_comparison.py (hunk 8 above)
Comment (umbernhard) [reply server/tests/ballot_comparison/snapshots/snap_test_ballot_comparison.py]: Yeah. When we changed the report to just replace any values less than 0.00000001 with that string, we didn't special case p-values of 0. I highly doubt we'll ever be in this situation though.

**Review by jonahkagan — APPROVED** [review]
